File: SignalServiceKit/tests/Messages/OWSDisappearingMessagesJobTest.m
//
//  Copyright (c) 2020 Open Whisper Systems. All rights reserved.
//

#import "OWSDisappearingMessagesConfiguration.h"
#import "OWSDisappearingMessagesFinder.h"
#import "OWSDisappearingMessagesJob.h"
#import "SSKBaseTestObjC.h"
#import "TSContactThread.h"
#import "TSIncomingMessage.h"
#import "TSMessage.h"
#import <SignalCoreKit/NSDate+OWS.h>
#import <SignalServiceKit/SignalServiceKit-Swift.h>

NS_ASSUME_NONNULL_BEGIN

@interface OWSDisappearingMessagesJobTest : SSKBaseTestObjC

@property TSThread *thread;

@end

@implementation OWSDisappearingMessagesJobTest

- (SignalServiceAddress *)localAddress
{
    return [[SignalServiceAddress alloc] initWithPhoneNumber:@"+13334445555"];
}

- (SignalServiceAddress *)otherAddress
{
    return [[SignalServiceAddress alloc] initWithPhoneNumber:@"+12223334444"];
}

- (void)setUp
{
    [super setUp];

    self.thread = [TSContactThread getOrCreateThreadWithContactAddress:self.localAddress];
}

- (TSMessage *)messageWithBody:(NSString *)body
              expiresInSeconds:(uint32_t)expiresInSeconds
               expireStartedAt:(uint64_t)expireStartedAt
{
    TSIncomingMessageBuilder *incomingMessageBuilder =
        [TSIncomingMessageBuilder incomingMessageBuilderWithThread:self.thread messageBody:body];
    incomingMessageBuilder.timestamp = 1;
    incomingMessageBuilder.expiresInSeconds = expiresInSeconds;
    incomingMessageBuilder.expireStartedAt = expireStartedAt;
    return [incomingMessageBuilder build];
}

- (void)testRemoveAnyExpiredMessage
{
    uint64_t now = [NSDate ows_millisecondTimeStamp];
    TSMessage *expiredMessage1 =
        [self messageWithBody:@"expiredMessage1" expiresInSeconds:1 expireStartedAt:now - 20000];

    TSMessage *expiredMessage2 =
        [self messageWithBody:@"expiredMessage2" expiresInSeconds:2 expireStartedAt:now - 2001];

    TSMessage *notYetExpiredMessage =
        [self messageWithBody:@"notYetExpiredMessage" expiresInSeconds:20 expireStartedAt:now - 10000];

    TSMessage *unExpiringMessage = [self messageWithBody:@"unexpiringMessage" expiresInSeconds:0 expireStartedAt:0];

    [self writeWithBlock:^(SDSAnyWriteTransaction *transaction) {
        [expiredMessage1 anyInsertWithTransaction:transaction];
        [expiredMessage2 anyInsertWithTransaction:transaction];
        [notYetExpiredMessage anyInsertWithTransaction:transaction];
        [unExpiringMessage anyInsertWithTransaction:transaction];
    }];

    OWSDisappearingMessagesJob *job = [OWSDisappearingMessagesJob sharedJob];

    // Sanity Check.
    [self readWithBlock:^(SDSAnyReadTransaction *transaction) {
        XCTAssertEqual(4, [TSMessage anyCountWithTransaction:transaction]);
    }];
    [job syncPassForTests];
    [self readWithBlock:^(SDSAnyReadTransaction *transaction) {
        XCTAssertEqual(2, [TSMessage anyCountWithTransaction:transaction]);
    }];
}

@end

NS_ASSUME_NONNULL_END

File: SignalServiceKit/tests/Messages/OWSDisappearingMessagesJobTest.swift
//
// Copyright 2022 Signal Messenger, LLC
// SPDX-License-Identifier: AGPL-3.0-only
//

import XCTest
import SignalServiceKit

final class OWSDisappearingMessagesJobTest: SSKBaseTestSwift {
    private func message(
        withBody body: String,
        expiresInSeconds: UInt32,
        expireStartedAt: UInt64
    ) -> TSMessage {
        let localAddress = SignalServiceAddress(uuid: UUID())
        let thread = TSContactThread.getOrCreateThread(contactAddress: localAddress)
        let messageBuilder = TSIncomingMessageBuilder.incomingMessageBuilder(
            thread: thread,
            messageBody: body
        )
        messageBuilder.timestamp = 1234
        messageBuilder.expiresInSeconds = expiresInSeconds
        messageBuilder.expireStartedAt = expireStartedAt
        return messageBuilder.build()
    }

    func testRemoveAnyExpiredMessage() {
        let now = Date.ows_millisecondTimestamp()
        let expiredMessage1 = message(
            withBody: "expired message 1",
            expiresInSeconds: 1,
            expireStartedAt: now - 20_000
        )
        let expiredMessage2 = message(
            withBody: "expired message 2",
            expiresInSeconds: 2,
            expireStartedAt: now - 2001
        )
        let notYetExpiredMessage = message(
            withBody: "not yet expired",
            expiresInSeconds: 20,
            expireStartedAt: now - 10_000
        )
        let messageThatDoesntExpire = message(
            withBody: "message that doesn't expire",
            expiresInSeconds: 0,
            expireStartedAt: 0
        )

        write { transaction in
            expiredMessage1.anyInsert(transaction: transaction)
            expiredMessage2.anyInsert(transaction: transaction)
            notYetExpiredMessage.anyInsert(transaction: transaction)
            messageThatDoesntExpire.anyInsert(transaction: transaction)
        }

        let job = OWSDisappearingMessagesJob.shared

        read { transaction in
            let messageCount = TSMessage.anyCount(transaction: transaction)
            XCTAssertEqual(messageCount, 4, "Test is not set up correctly")
        }

        job.syncPassForTests()

        read { transaction in
            let messageCount = TSMessage.anyCount(transaction: transaction)
            XCTAssertEqual(messageCount, 2)
        }
    }
}
